dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp

```diff
@@ -99,32 +107,58 @@ std::pair<ASTTableJoin::Kind, size_t> getJoinKindAndBuildSideIndex(const tipb::J
+    }
+    catch (DB::Exception & e)
+    {
+        if (e.code() == ErrorCodes::NO_COMMON_TYPE
+            && removeNullable(left_type)->isDecimal()
+            && removeNullable(right_type)->isDecimal())
+        {
+            // fix https://github.com/pingcap/tiflash/issues/4519
+            // String is the common type for all types, it is always safe to choose String.
+            // But then we need to use `FunctionFormatDecimal` to format decimal.
+            // Such as 0.1000000000 and 0.10000000000000000000.
+            RUNTIME_ASSERT(!left_type->onlyNull() || !right_type->onlyNull());
+            auto fall_back_type = std::make_shared<DataTypeString>();
+            bool make_nullable = left_type->isNullable() || right_type->isNullable();
+            return {make_nullable ? makeNullable(fall_back_type) : fall_back_type, true};
+        }
+        else
+        {
+            throw;
+        }
+    }
+}
+
+JoinKeyTypes getJoinKeyTypes(const tipb::Join & join)
 {
     if (unlikely(join.left_join_keys_size() != join.right_join_keys_size()))
         throw TiFlashException("size of join.left_join_keys != size of join.right_join_keys", Errors::Coprocessor::BadRequest);
-    DataTypes key_types;
+    JoinKeyTypes join_key_types;
     for (int i = 0; i < join.left_join_keys_size(); ++i)
     {
         if (unlikely(!exprHasValidFieldType(join.left_join_keys(i)) || !exprHasValidFieldType(join.right_join_keys(i))))
             throw TiFlashException("Join key without field type", Errors::Coprocessor::BadRequest);
-        DataTypes types;
-        types.emplace_back(getDataTypeByFieldTypeForComputingLayer(join.left_join_keys(i).field_type()));
-        types.emplace_back(getDataTypeByFieldTypeForComputingLayer(join.right_join_keys(i).field_type()));
-        DataTypePtr common_type = getLeastSupertype(types);
-        key_types.emplace_back(common_type);
+        auto left_type = getDataTypeByFieldTypeForComputingLayer(join.left_join_keys(i).field_type());
+        auto right_type = getDataTypeByFieldTypeForComputingLayer(join.right_join_keys(i).field_type());
+        join_key_types.emplace_back(geCommonTypeForJoinOn(left_type, right_type));
     }
-    return key_types;
+    return join_key_types;
 }
 
-TiDB::TiDBCollators getJoinKeyCollators(const tipb::Join & join, const DataTypes & join_key_types)
+TiDB::TiDBCollators getJoinKeyCollators(const tipb::Join & join, const JoinKeyTypes & join_key_types)
 {
     TiDB::TiDBCollators collators;
     size_t join_key_size = join_key_types.size();
     if (join.probe_types_size() == static_cast<int>(join_key_size) && join.build_types_size() == join.probe_types_size())
         for (size_t i = 0; i < join_key_size; ++i)
         {
-            if (removeNullable(join_key_types[i])->isString())
+            if (removeNullable(join_key_types[i].key_type)->isString())
             {
                 if (unlikely(join.probe_types(i).collate() != join.build_types(i).collate()))
```

Comment:
Don't need to check the collate if the original type is Decimal

Reply:
ok, updated.